tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpIngressTimeTest.java

```diff
@@ -0,0 +1,98 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.activemq.artemis.tests.integration.amqp;
+
+import java.util.concurrent.TimeUnit;
+
+import org.apache.activemq.artemis.core.server.Queue;
+import org.apache.activemq.artemis.core.settings.impl.AddressSettings;
+import org.apache.activemq.artemis.protocol.amqp.converter.AMQPMessageSupport;
+import org.apache.activemq.transport.amqp.client.AmqpClient;
+import org.apache.activemq.transport.amqp.client.AmqpConnection;
+import org.apache.activemq.transport.amqp.client.AmqpMessage;
+import org.apache.activemq.transport.amqp.client.AmqpReceiver;
+import org.apache.activemq.transport.amqp.client.AmqpSession;
+import org.junit.Test;
+
+public class AmqpIngressTimeTest extends AmqpClientTestSupport {
+
+   @Test(timeout = 60000)
+   public void testIngressTimestampSendCore() throws Exception {
+      internalTestIngressTimestamp(Protocol.CORE);
+   }
+
+   @Test(timeout = 60000)
+   public void testIngressTimestampSendAMQP() throws Exception {
+      internalTestIngressTimestamp(Protocol.AMQP);
+   }
+
+   @Test(timeout = 60000)
+   public void testIngressTimestampSendOpenWire() throws Exception {
+      internalTestIngressTimestamp(Protocol.OPENWIRE);
+   }
+
+   private void internalTestIngressTimestamp(Protocol protocol) throws Exception {
+      server.getAddressSettingsRepository().addMatch(getQueueName(), new AddressSettings().setEnableIngressTime(true));
+      long beforeSend = System.currentTimeMillis();
+      if (protocol == Protocol.CORE) {
+         sendMessagesCore(getQueueName(), 1, true);
+      } else if (protocol == Protocol.OPENWIRE) {
+         sendMessagesOpenWire(getQueueName(), 1, true);
+      } else {
+         sendMessages(getQueueName(), 1, true);
+      }
+      long afterSend = System.currentTimeMillis();
+
+      server.stop();
+      server.start();
+      assertTrue(server.waitForActivation(3, TimeUnit.SECONDS));
+
+      AmqpClient client = createAmqpClient();
+      AmqpConnection connection = addConnection(client.connect());
+      AmqpSession session = connection.createSession();
+
+      AmqpReceiver receiver = session.createReceiver(getQueueName());
+
+      Queue queueView = getProxyToQueue(getQueueName());
+      assertEquals(1, queueView.getMessageCount());
+
+      receiver.flow(1);
+      AmqpMessage receive = receiver.receive(5, TimeUnit.SECONDS);
+      assertNotNull(receive);
+      instanceLog.info(receive);
+      Object ingressTimestamp = receive.getMessageAnnotation(AMQPMessageSupport.HDR_INGRESS_TIME);
+      assertNotNull(ingressTimestamp);
+      assertTrue(ingressTimestamp instanceof Long);
+      long ingressTs = (Long) ingressTimestamp;
+      assertTrue("Ingress time should be set in expected range",beforeSend <= ingressTs && afterSend >= ingressTs);
```

Comment:
Same point as last time about the unhelpful failure this will produce

Reply:
Are you talking about the assertions that have no error message or the specific message on the assertion which does?

Reply:
Im not sure why GitHub is showing 4 lines as I only selected one at the time I commented, but its the last assertTrue with the message I mean, i.e "assertTrue("Ingress time should be set in expected range",beforeSend <= ingressTs && afterSend >= ingressTs);"

The above will basically give 'it failed here, good luck on your quest', basically requiring you run it again to try and figure out what happened (which could be racey and hard to reproduce, might cause us to modify the test to help figure the reason, etc), whereas a more useful failure would show what the value was and what it should have been [between]. That might even allow figuring out the issue without needing to run the test again (or if we cant get it to fail again) as the failure message was far more useful and potential causes potentially could be reasoned from it...e.g 'oh it was 0, weve stopped setting it, oops \<checks recent changes\>' or 'it was 1ms outside expectation, possibly we introduced some unexpected async behaviour'

E.g one option from one of the earlier comments: https://github.com/apache/activemq-artemis/pull/3279#discussion_r533325701
```
Matcher<Long> inRange = both(greaterThanOrEqualTo(beforeSend)).and(lessThanOrEqualTo(afterSend));
MatcherAssert.assertThat("Ingress time should be set in expected range", ingressTs, inRange);
```

Reply:
Fixed.